### PR TITLE
fix: retain cycle when removing devices [WPB-6953]

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6953" title="WPB-6953" target="_blank">BugWPB-6953</a>  Multiple alerts for removing clients are shown
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

A retain cycle led to observations not being torn down and that led to the password input alert dialog being presented more than once when removing self clients.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+Clients.swift [hunk 1]
```diff
@@ -66,24 +66,24 @@ extension ZMUserSession {
     /// - Returns: Token that needs to be stored as long the observer should be active.
 
     @objc(addClientUpdateObserver:)
-    public func addClientUpdateObserver(_ observer: ClientUpdateObserver) -> Any {
+    public func addClientUpdateObserver(_ observer: ClientUpdateObserver) -> NSObjectProtocol {
 
-        return ZMClientUpdateNotification.addObserver(context: managedObjectContext) { [weak self] (type, clientObjectIDs, error) in
+        return ZMClientUpdateNotification.addObserver(context: managedObjectContext) { [weak self, weak observer] type, clientObjectIDs, error in
```
Comment (netbe) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+Clients.swift:71]: do we need to patch also all other observers ?

Comment (caldrian) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+Clients.swift:71]: I'm sure we have many more memory management issues, but in this case it led to wrong behavior and fixing it was the scope of this ticket and PR.

**Review by netbe — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift [hunk 2]
```diff
@@ -69,26 +69,22 @@ final class DeviceDetailsViewActionsHandler: DeviceDetailsViewActions, Observabl
...
-            guard let self = self else {
-                return
+            guard let self else {
+                return continuation.resume(returning: false)
             }
-            // (Continuation)[https://developer.apple.com/documentation/swift/checkedcontinuation]
-            // Using the same continuation twice results in a crash.
-            var optionalContinuation: CheckedContinuation<Bool, Never>? = continuation
+
             clientRemovalObserver = ClientRemovalObserver(
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift:76]: glad this hack is gone

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+Clients.swift (hunk 1 above)
Comment (caldrian) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+Clients.swift:71]: Short explanation: the observer that was passed into this method was retained by the closure. The token, which kept the observation alive, is stored in the very same instance.

**Review by typfel — APPROVED** [review]

**Pull request merged**